rasa/core/processor.py

```diff
@@ -374,8 +376,33 @@ def predict_next_action(
         """Predicts the next action the bot should take after seeing x.
 
         This should be overwritten by more advanced policies to use
-        ML to predict the action. Returns the index of the next action.
+        ML to predict the action.
+
+        Returns:
+             The index of the next action and prediction of the policy.
+
+        Raises:
+            ActionLimitReached if the limit of actions to predict has been reached.
         """
+        should_predict_another_action = self.should_predict_another_action(
+            tracker.latest_action_name
+        )
+        reversed_events = list(tracker.events)[::-1]
+        num_predicted_actions = 0
+
+        for e in reversed_events:
+            if isinstance(e, ActionExecuted):
+                if e.action_name in (ACTION_LISTEN_NAME, ACTION_SESSION_START_NAME):
+                    break
+                num_predicted_actions += 1
```

Comment:
Isn't this duplicate logic from what's in `_predict_and_execute_next_action`?

Reply:
It looks similar but not quite - the while loop in `_predict_and_execute_next_action` also has a check for `_should_handle_message` which `predict_next_action` doesn't need. 

I did ask Alex[ if I should also adapt this method ](https://github.com/RasaHQ/rasa/pull/9173#issuecomment-883415430) to reduce duplicate logic from `predict_next_action` however[ it seems out of scope for this issue](https://github.com/RasaHQ/rasa/pull/9173#issuecomment-883936445).

Reply:
but since `_predict_and_execute_next_action` is calling `predict_next_action` we don't want the same things to be done twice right? Also we are now making changes to the processor, not just the test module, so we need to think about the broader implications.

Reply:
I'm happy to submit a different issue for this specific processor method - the original issue for this PR was focused on making sure `rasa.core.test` doesn't duplicate/expose logic from the processor. 
Otherwise, happy to accept suggestions how to adapt `processor._predict_and_execute_next_action`.

Reply:
Ok, method adapted 🎉  I tried first with raising a new exception `if not should_predict_another_action`, but that broke multiple unit tests in both `test_processor` and `test_evaluation` 🙈  So I arrived at the current state. Let me know what you think.